### PR TITLE
[move-lang] move the shadow compilation logic from move-cli to source compiler

### PR DESCRIPTION
**Pull request by mengxu-fb**
## Motivation

The feature to allow Move module compilation with duplicating/conflicting definitions against library modules is experimented in Move CLI. But ideally, this feature should live in the source compiler. This PR moves the logic there.

It also simplifies the filtering logic by only looking at the `parser::ast::Program` instead of going through the `mv_interfaces` directory with filename matching.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

`cargo xtest`

### CODE REVIEW

File: language/move-lang/src/lib.rs
Comment (sblackshear) [inline language/move-lang/src/lib.rs:183]: A comment here would be useful

File: language/move-lang/src/lib.rs
Comment (sblackshear) [inline language/move-lang/src/lib.rs:385]: Same here

**Review by sblackshear — APPROVED** [review]
Nice!

**Comment by mengxu-fb** [comment]
Comments added, would like @tnowacki to have a final look before I land it :)

File: language/move-lang/src/lib.rs
Comment (tnowacki) [inline language/move-lang/src/lib.rs:217]: I just went through a big refactor trying to slim this down, so I'm not really keen on adding another entry point. 

I think `sources_shadow_deps` should just be a flag on all compiler entry points

**Review by tnowacki — CHANGES_REQUESTED** [review]
Looks simple enough! I'm just trying to keep things slim, so let's see how passing the flag in works

File: language/move-lang/src/lib.rs
Comment (tzakian) [inline language/move-lang/src/lib.rs:198]: Maybe a larger question, but I'm wondering why we are using two different versions of `Result` here...It seems like other places in this file are doing a similar thing, so this doesn't block this PR, but it might be useful to mark this as a future cleanup unless there's a good reason for this. @tnowacki curious if you think this cleanup would make sense.

Comment (mengxu-fb) [reply language/move-lang/src/lib.rs:198]: In fact, I had the same wondering here :)

File: language/move-lang/src/lib.rs
Comment (mengxu-fb) [reply language/move-lang/src/lib.rs:217]: Just a heads up, this will lead to many changes across the `language/` dir if every entry points of the compiler need to be updated...

File: language/move-lang/src/bin/move-check.rs
Comment (tnowacki) [inline language/move-lang/src/bin/move-check.rs:55]: can you set the default to be true? 

I feel like it is more intuitive than the alternative. But thoughts?

Comment (mengxu-fb) [reply language/move-lang/src/bin/move-check.rs:55]: I don't have a strong opinion either way.

Comment (mengxu-fb) [reply language/move-lang/src/bin/move-check.rs]: Made an update to default it to be true.

File: language/move-lang/src/lib.rs
Comment (tnowacki) [reply language/move-lang/src/lib.rs]: Its tricky....

The outer result is for random IO nonsense, the inner one is for errors from the compiler itself. 
IF you wanted to clean this up, you could handroll an enum for the 3 states. It would be clearer but I'm not sure if it would be simpler

**Review by tnowacki — APPROVED** [review]
LGTM. Thanks for doing this!

**Pull request closed without merging**
